protocols/ping/src/handler.rs

```diff
@@ -184,6 +184,18 @@ impl Handler {
     ) {
         self.outbound = None; // Request a new substream on the next `poll`.
 
+        // Timer is already polled and expired before substream request is initiated
+        // and will be polled again later on in our `poll` because we reset `self.outbound`.
+        //
+        // `futures-timer` allows an expired timer to be polled again and returns
+        // immidietly a `Poll::Ready`. However in its WASM implementation there is
+        // a bug that causes the expired timer to panic.
+        //
+        // This is a workaround until a proper fix is merged and released.
+        //
+        // See async-rs/futures-timer#74 and libp2p/rust-libp2p#5447 for more info.
+        self.interval.reset(Duration::new(0, 0));
```

Comment:
in the two instances that I see `self.outbound` be set to `None`, here (line `185`) and on `Handler::new`  `self.interval` is `reset` with `Duration::new(0, 0)`, which means that we don't care about `self.interval` when `self.outbound` is `None` right? If so, it makes more sense to just skip polling `self.interval` here:
https://github.com/libp2p/rust-libp2p/blob/3b61e57d0e72d371ccdfc134bd71cb3aa20addee/protocols/ping/src/handler.rs#L310-L318

Reply:
You missed one case: Line 289 also sets `outbound`  to `None` and in the case of a `OutboundState::Ping` and later on `Poll::Ready(Err(_))` the timer is set to the ping interval.

The question is: Since the retry of this case is delayed, should the dial error follow the same way and delay the retry?

https://github.com/oblique/rust-libp2p/blob/2ea753ceafdd289e7934394fc044165912e9d4e6/protocols/ping/src/handler.rs#L289-L306

Reply:
Thanks for expanding Yianis, I don't think the dial upgrade error should follow, as you stated it wasn't the previous behaviour.